src/aws/flb_aws_util.c

```diff
@@ -322,15 +328,46 @@ struct flb_http_client *request_do(struct flb_aws_client *aws_client,
         flb_warn("[aws_http_client] failed to increase max response buffer size");
     } 
 
-    /* Add AWS Fluent Bit user agent */
-    if (aws_client->extra_user_agent == NULL) {
+    /* 
+    * Set AWS Fluent Bit user agent FLB_AWS_USER_AGENT
+    * This environment variable needs to be set when:
+    * 1. it is not set;
+    * 2. extra_user_agent has been specified by one aws_client and 
+    * its value doesn't match current AWS Fluent Bit user agent
+    */
+    env = aws_client->upstream->config->env;
+    buf = (char *) flb_env_get(env, "FLB_AWS_USER_AGENT");
+    if (buf == NULL || 
+       (aws_client->extra_user_agent != NULL && strcasecmp(buf, aws_client->extra_user_agent) != 0)) {
```

Comment:
the logic here seems different than what we discussed last week. I thought we agreed that extra_user_agent is most specific since its per plugin and so it takes precedence. Hence if its set, this code block shoulnd't run and the extra_user_agent should just be used?

Reply:
yeah so the reason why I implement like this is in my last comment. I tested more after our discussion and updated it. And in current logic, I think extra_user_agent is also our highest priority setting/case to set

Reply:
I see. I misunderstood the code, but this is still not the logic I thought we had agreed on. See I thought we agreed extra_user_agent was plugin specific. Which means it only affects a single instance of the client. So we should not set the extra_user_agent value with `flb_env_set` because then we've made an instance config a global setting. 

> yeah so the problem is that based on my test, even we set extra_user_agent in CW config, it is not set at the very beginning. Fluent Bit will call this file first (without CW client triggered) and then trigger CW (with extra_user_agent set then)

This should be the case. Say we are running in EKS and we have three outputs which are registered in this order:

1. kinesis_firehose
2. cloudwatch_logs with extra_user_agent set
3. cloudwatch_logs without extra_user_agent set

So in the AWS Client code, the registration should look like this:
1. Create client for kinesis_firehose, there is no extra_user_agent for it and the FLB_AWS_USER_AGENT is not set in the env yet so we find that we are in k8s and then set FLB_AWS_USER_AGENT=k8s.
2. Since extra_user_agent is set for this client/plugin, it just uses that and does not check or set FLB_AWS_USER_AGENT at all. 
3. The final plugin runs its init and this time when the client is created it should see that it has no extra_user_agent but that FLB_AWS_USER_AGENT is already set, hence it will use that.